In [1]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

## 1. Understand Table Structure

*Checking the available columns and their data types*

In [4]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'return_data';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


column_name,data_type
value_return,double precision
qty_returned,integer
return_date,date
sku_id,character varying
nama_produk,character varying
status_return,character varying
return_reason,character varying
retur_id,character varying
invoice_id,character varying
so_id,character varying


**Findings**
- `value_return`typed as `float` → akan digunakan untuk menghitung total nilai kerugian akibat retur.
- `return_date` typed as `date` → kalkulasi tren bulanan dapat dilakukan langsung tanpa konversi.
- `nama_produk` and `sku_id` typed as `varchar` → `sku_id` akan digunakan sebagai foreign key untuk JOIN ke `master_sku` guna mendapatkan kategori produk.
- `return_reason` typed as `varchar` → akan digunakan untuk menganalisis alasan retur paling sering.
- `invoice_id` typed as `varchar` → akan digunakan sebagai foreign key untuk JOIN ke `invoice_data` guna mendapatkan informasi customer.

In [5]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'master_sku';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
price_max,double precision
price_min,double precision
kategori,character varying
sku_id,character varying
sku_status,character varying
satuan,character varying
sku_name,character varying


**Findings**
- `sku_id` typed as `varchar` →  primary key tabel ini, digunakan sebagai jembatan JOIN dari `return_data` ke `master_sku`
- `kategory`typed as `varchar` → akan digunakan untuk menganalisis kategory yang paling sering retur.
- `sku_name`typed as `varchar` → akan digunakan untuk menganalisis produk yang paling sering retur.

In [6]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'invoice_data';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
11 rows affected.


column_name,data_type
has_return,boolean
invoice_date,date
is_partial,boolean
partial_ke,double precision
invoice_value,double precision
bundle_complete_date,date
so_id,character varying
customer_id,character varying
invoice_id,character varying
payment_status,character varying


**Findings**
- `customer_id` typed as `varchar` →  primary key tabel ini, digunakan sebagai jembatan JOIN dari `return_data` ke `master_customer`
- `invoice_id` typed as `varchar` → akan digunakan sebagai foreign key untuk JOIN ke `return_data`

In [7]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'master_customer';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
customer_id,character varying
customer_name,character varying
customer_type,character varying
region,character varying
alamat,character varying
payment_terms,character varying
customer_status,character varying


**Findings**
- `customer_id` typed as `varchar` → primary key tabel ini, diakses melalui JOIN
- `region` typed as `varchar` → akan digunakan untuk menganalisis region yang paling sering retur.
- `customer_type` typed as `varchar` → akan digunakan untuk menganalisis tipe customer yang paling sering retur.

**Conclusion:**  
Semua tabel memiliki struktur yang valid. Foreign key antar tabel 
tersedia dan konsisten untuk mendukung JOIN di tahap analisis.

## 2. Data Preview

*Checking sample records to verify the data makes sense*

In [8]:
%%sql

SELECT *
FROM
    return_data
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


retur_id,invoice_id,so_id,sku_id,nama_produk,qty_returned,return_reason,return_date,status_return,value_return
RT00001,009502,SO240001,B006,Klem Arteri Bengkok,9,Salah Pesan Customer,2024-10-22,In Progress,1440000.0
RT00002,009510,SO240007,B014,Duk Operasi 80x90cm,22,Salah Pesan Customer,2024-07-21,Completed,1067000.0
RT00003,009515,SO240011,C019,NGT 16Fr,31,Kadaluarsa,2024-02-22,Completed,899000.0
RT00004,009541,SO240029,C006,Infus Set Anak,14,Salah Pesan Customer,2024-10-12,Completed,182000.0
RT00005,009543,SO240031,L010,Objek Glass Frosted,10,Salah Input Sistem,2024-08-22,Completed,855000.0


**Findings**
- Data terlihat valid dan masuk akal — `return_date` dalam format yang konsisten, `value_return` berisi nilai numerik yang wajar
- `return_reason` terlihat berisi nilai kategorikal — dari sample ini muncul Salah Pesan Customer, Kadaluarsa, dan Salah Input Sistem
- `status_return` memiliki dua nilai yang terlihat di sample ini: Completed dan In Progress


In [13]:
%%sql

SELECT *
FROM
    master_sku
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


sku_id,sku_name,kategori,satuan,price_min,price_max,sku_status
C001,Masker Bedah 3-Ply,Consumable,Box,35000.0,55000.0,Active
C002,Sarung Tangan Nitril S,Consumable,Box,120000.0,180000.0,Active
C003,Sarung Tangan Nitril M,Consumable,Box,120000.0,180000.0,Active
C004,Sarung Tangan Nitril L,Consumable,Box,120000.0,180000.0,Active
C005,Infus Set Dewasa,Consumable,Pcs,8000.0,15000.0,Active


**Findings**
- `sku_name` terlihat valid — nama produk terbaca jelas dan konsisten
- `kategori` dari sample ini hanya terlihat 1 kategori (`consumable`), namun pengecekan nilai unik akan dilakukan di section selanjutnya


In [10]:
%%sql

SELECT *
FROM
    invoice_data
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


invoice_id,so_id,customer_id,invoice_date,is_partial,partial_ke,invoice_value,payment_status,status_invoice,bundle_complete_date,has_return
009723,SO240166,KLN068,2024-09-16,True,1.0,27138500.0,NET 30,Issued,None,False
010054,SO240404,KLN003,2024-02-12,True,1.0,6318000.0,NET 30,Bundle Complete,2024-03-07,False
009762,SO240195,RSS045,2024-02-23,True,1.0,21065000.0,NET 30,Bundle Complete,2024-03-20,False
009808,SO240226,RSP024,2024-11-18,True,2.0,14670000.0,NET 45,Issued,None,False
009926,SO240312,RSS066,2024-08-10,True,2.0,28354000.0,NET 30,Bundle Complete,2024-09-14,True


**Findings**
- `invoice_id` dan `customer_id` terlihat valid dan konsisten — siap digunakan sebagai jembatan JOIN dari `return_data` ke `master_customer`


In [11]:
%%sql

SELECT *
FROM
    master_customer
LIMIT 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


customer_id,customer_name,customer_type,region,alamat,payment_terms,customer_status
RSP001,RSUD Zulaika,RS Pemerintah,Jakarta Barat,"Jl. KH Amin Jasuta No. 173, Cirebon, JA 74749",NET 45,Active
RSS001,RS Suryono,RS Swasta,Tangerang,"Gang Dipatiukur No. 1, Tebingtinggi, Kepulauan Riau 46986",NET 30,Active
KLN001,Klinik Laksita,Klinik,Jakarta Timur,"Gang Surapati No. 0, Payakumbuh, Bali 83302",NET 30,Active
RSS002,RS Wasita,RS Swasta,Jakarta Utara,"Gg. Surapati No. 10, Surakarta, Maluku Utara 15272",NET 30,Active
KLN002,Klinik Sihombing,Klinik,Jakarta Selatan,"Jalan Rajawali Timur No. 088, Palangkaraya, BA 39649",NET 30,Active


**Findings**
- `region` terlihat valid — nama region terbaca jelas, namun pengecekan inkonsisten akan dilakukan di section selanjutnya
- `customer_type` terlihat berisi nilai kategorikal — dari sample ini muncul RS Pemerintah, RS Swasta, dan Klinik, namun pengecekan inkonsisten akan dilakukan di section selanjutnya


**Conclusion:**  
Data preview dari keempat tabel terlihat valid dan masuk akal. Potensi inkonsistensi penulisan ditemukan pada kolom `region` dan `customer_type` di `master_customer` — akan diverifikasi lebih lanjut di section Check Unique Values.

## 3. Count Total Records

*Checking the total number of rows in each table*

In [14]:
%%sql

SELECT COUNT(*) as total_records
FROM
    return_data;


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
59


In [15]:
%%sql

SELECT COUNT(*) as total_records
FROM
    master_sku;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
73


In [16]:
%%sql

SELECT COUNT(*) as total_records
FROM
    invoice_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
703


In [17]:
%%sql

SELECT COUNT(*) as total_records
FROM
    master_customer;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
220


**Findings**
- `return_data` memiliki 59 records — ini adalah total transaksi retur yang akan dianalisis sepanjang 2024
- `master_sku` memiliki 73 produk yang terdaftar
- `master_customer` memiliki 220 instansi kesehatan yang terdaftar
- `invoice_data` memiliki 703 records — jauh lebih banyak dari `return_data`, artinya tidak semua invoice berujung pada retur


**Conclusion:**  
Dataset cukup untuk dianalisis. Dari 703 invoice yang tercatat, hanya 59 yang menghasilkan retur — mengindikasikan retur tidak terjadi di semua transaksi.

## 4. Check Missing Values

*Checking for missing values in key columns*

In [22]:
%%sql

SELECT
    COUNT(*) - COUNT(value_return) AS missing_value_return,
    COUNT(*) - COUNT(return_date) AS missing_return_date,
    COUNT(*) - COUNT(nama_produk) AS missing_nama_product,
    COUNT(*) - COUNT(sku_id) AS missing_sku_id,
    COUNT(*) - COUNT(invoice_id) AS missing_invoice_id,
    COUNT(*) - COUNT(return_reason) AS missing_return_reason
FROM
    return_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_value_return,missing_return_date,missing_nama_product,missing_sku_id,missing_invoice_id,missing_return_reason
0,0,0,0,0,0


In [24]:
%%sql

SELECT
    COUNT(*) - COUNT(kategori) AS missing_kategory,
    COUNT(*) - COUNT(sku_name) AS missing_sku_name
FROM
    master_sku;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_kategory,missing_sku_name
0,0


In [27]:
%%sql

SELECT
    COUNT(*) - COUNT(customer_id) AS missing_customer_id
FROM
    invoice_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_customer_id
0


In [28]:
%%sql

SELECT
    COUNT(*) - COUNT(region) AS missing_region,
    COUNT(*) - COUNT(customer_name) AS missing_customer_name
FROM
    master_customer;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_region,missing_customer_name
0,0


**Findings**  
Tidak ditemukan missing values pada seluruh kolom yang relevan di keempat tabel — `return_data`, `master_sku`, `invoice_data`, dan `master_customer`

**Conclusion:**  
Semua tabel bersih dari missing values dan siap untuk tahap analisis selanjutnya.

## 5. Check for Duplicates

*Checking for records that appear more than once*

In [ ]:
%%sql

SELECT 
    retur_id, 
    COUNT(*) AS total_records
FROM 
    return_data
GROUP BY 
    retur_id
HAVING 
    COUNT(*) > 1;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


retur_id,jumlah


In [14]:
%%sql

SELECT
    sku_id,
    count(*) AS total_records
FROM
    master_sku
GROUP BY
    sku_id
HAVING
    count(*) > 1

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


sku_id,total_records


In [16]:
%%sql

SELECT
    invoice_id,
    count(*) AS total_records
FROM
    invoice_data
GROUP BY
    invoice_id
HAVING
    COUNT(*)>1

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
20 rows affected.


invoice_id,total_records
009918,2
009902,2
010175,2
010129,2
009514,2
010063,2
009774,2
009986,2
009689,2
009873,2


In [22]:
%%sql
SELECT 
    *
FROM 
    invoice_data
WHERE 
    invoice_id = '009918';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
2 rows affected.


invoice_id,so_id,customer_id,invoice_date,is_partial,partial_ke,invoice_value,payment_status,status_invoice,bundle_complete_date,has_return
009918,SO240306,RSS036,2024-04-14,True,1.0,42586000.0,NET 30,Bundle Complete,2024-05-07,False
009918,SO240306,RSS036,2024-04-11,True,1.0,42586000.0,NET 30,Bundle Complete,2024-05-07,False


In [20]:
%%sql

SELECT
    customer_id,
    COUNT(*) AS total_records
FROM
    master_customer
GROUP BY
    customer_id
HAVING
    COUNT(*)>1

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


customer_id,total_records


**Finding:**  
- Tidak ditemukan duplikat pada `return_data`, `master_sku`, dan `master_customer`.
- Ditemukan duplikat pada `invoice_data` — beberapa `invoice_id` muncul lebih dari satu kali dengan seluruh kolom yang identik, mengindikasikan data ter-input dua kali secara tidak sengaja (bukan disebabkan oleh partial invoice)

**Conclusion:**  
Duplikat pada `invoice_data` perlu di-drop di tahap Data Cleaning sebelum JOIN dilakukan, karena dapat menyebabkan inflasi jumlah baris pada hasil analisis.

## 6. Check Cross-Table Consistency

*Verifying that all foreign keys in `return_data` are properly registered in their respective reference tables — `sku_id` in `master_sku`, `invoice_id` in `invoice_data`, and `customer_id` (via `invoice_data`) in `master_customer`.*

In [26]:
%%sql

SELECT
    COUNT (*) AS unregistered_sku
FROM
    return_data
WHERE
    sku_id NOT IN
        (
            SELECT
                sku_id
            FROM
                master_sku
        )


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


unregistered_sku
0


In [27]:
%%sql

SELECT
    COUNT(*) AS unregistered_invoice
FROM
    return_data
WHERE
    invoice_id NOT IN
        (
            SELECT
                invoice_id
            FROM
                invoice_data
        )

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


unregistered_invoice
0


In [30]:
%%sql

SELECT
    COUNT (*) AS unregistered_customer
FROM
    invoice_data
WHERE
    customer_id NOT IN
        (
            SELECT
                customer_id
            FROM
                master_customer
        )

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


unregistered_customer
0


**Findings:**
Tidak ditemukan `sku_id`, `invoice_id`, maupun `customer_id` yang tidak terdaftar di tabel referensinya masing-masing — seluruh foreign key di `return_data` dan `invoice_data` valid dan dapat di-JOIN dengan aman.

**Conclusion**  
Referential integrity antar tabel terjaga dengan baik. Tidak ada risiko kehilangan data saat proses JOIN di tahap analisis selanjutnya.


## 7. Verify Critical Data Types

*Confirming that `return_date` and `value_return` are correctly typed as DATE and numeric, not text.*

In [38]:
%%sql

SELECT
    column_name, 
    data_type
FROM
    information_schema.columns
WHERE
    table_name = 'return_data'
    AND column_name IN ('return_date', 'value_return')

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
2 rows affected.


column_name,data_type
return_date,date
value_return,double precision


**Findings:**
- `return_date` → DATE 
- `value_return` → double precision 

**Conclusion**
Kedua kolom kritis sudah bertipe data yang sesuai — kalkulasi tren bulanan pada `return_date` dan penjumlahan nilai kerugian pada `value_return` dapat dilakukan langsung tanpa konversi tambahan.

## 8. Check Unique Value

*Checking unique values in categorical columns to detect inconsistent formatting (e.g., mixed casing, abbreviations) before proceeding to Data Cleaning.*

In [43]:
%%sql

SELECT
    DISTINCT return_reason
    
FROM
    return_data

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
6 rows affected.


return_reason
Kadaluarsa
Salah Input Sistem
Salah Pesan Customer
customer salah pesan
Slh Input
Barang Rusak


**Findings:**  
`return_reason` ditemukan 6 variasi penulisan, namun beberapa merujuk pada alasan yang sama:
- `Salah Pesan Customer` = `customer salah pesan` → seharusnya satu kategori
- `Salah Input Sistem` = `Slh Input` → seharusnya satu kategori
- `Kadaluarsa` dan `Barang Rusak` masing-masing berdiri sendiri sebagai kategori unik

In [46]:
%%sql

SELECT
    DISTINCT status_return
FROM
    return_data

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
2 rows affected.


status_return
Completed
In Progress


**Findings:**  
`status_return` hanya memiliki 2 nilai valid — `Completed` dan `In Progress`. Tidak ditemukan inkonsistensi penulisan.

In [47]:
%%sql

SELECT
    DISTINCT kategori
FROM
    master_sku

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori
Bedah & Tindakan
Diagnostik
Laboratorium
Consumable
Perawatan & Rawat Inap


**Findings:**  
`kategori` memiliki 5 nilai valid — `Bedah & Tindakan`, `Diagnostik`, `Laboratorium`, `Consumable` dan `Perawatan & Rawat Inap`. Tidak ditemukan inkonsistensi penulisan.

In [50]:
%%sql

SELECT
    DISTINCT customer_type
FROM
    master_customer

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
11 rows affected.


customer_type
Rumah Sakit Swasta
Klinik
RS Pemerintah
klinik
RS SWASTA
RSP
RS Swasta
rs swasta
Klinik Pratama
rs pemerintah


**Findings:**  
`customer_type` ditemukan 11 variasi penulisan, namun beberapa merujuk pada tipe customer yang sama:
- `RS Pemerintah` = `rs pemerintah`, dan `RSP` → seharusnya satu kategori
- `RS Swasta` = `Rumah Sakit Swasta`, `RS SWASTA`, `rs swasta`, dan `RSS` → seharusnya satu kategori
- `Klinik` = `klinik`, dan `Klinik Pratama` → seharusnya satu  kategori

In [51]:
%%sql

SELECT
    DISTINCT region
FROM
    master_customer

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
22 rows affected.


region
Jakarta Selatan
Jakarta Timur
Jkt Pusat
Bks
DEPOK
Dpk
JKT PUSAT
Jkt Selatan
bekasi
Jakarta Pusat


**Findings:**  
`region` ditemukan 22 variasi penulisan, namun beberapa merujuk pada region yang sama:
- `Jakarta Selatan` = `Jkt Selatan` → seharusnya satu kategori
- `Jakarta Barat` = `Jak-Bar`, dan `Jakarta barat` → seharusnya satu  kategori
- `Jakarta Pusat` = `Jkt Pusat`, dan `JKT PUSAT` → seharusnya satu  kategori
- `Bogor` = `bogor`, dan `BOGOR` → seharusnya satu  kategori
- `Depok` = `DEPOK`, dan `Dpk`,  → seharusnya satu  kategori
- `Tangerang` = `tangerang`, dan `TANGERANG` → seharusnya satu  kategori
- `Bekasi` = `Bks`, dan `bekasi` → seharusnya satu  kategori
- `Jakarta Utara` dan `Jakarta Timur` masing-masing berdiri sendiri dengan kategori unik

**Conclusion**  
Ditemukan inkonsistensi penulisan pada tiga kolom — `return_reason`, `customer_type`, dan `region` — sementara `status_return` dan `kategori` sudah konsisten. Standarisasi terhadap ketiga kolom yang bermasalah akan dilakukan di tahap Data Cleaning menggunakan logika `CASE WHEN` di SQL.